# "-fPIE" always triggers segmentation fault in ld.so.1 on mips platform,
# which was first saw on dbus-daemon. Below borrow the binutils fix from
# binutils bugzilla:
#
#	http://sourceware.org/bugzilla/show_bug.cgi?id=10858
#
# Its commit message says:
#	* elfxx-mips.c (mips_elf_create_dynamic_relocation): Use section
#	sym dynindx for relocs against defined syms in PIEs.
#
# It's in upstream CVS now (rev 1.267), but not in current release
#
# By Kevin Tian <kevin.tian@intel.com>, 2010-07-15

diff --git a/elfxx-mips.c b/elfxx-mips.c
index 3a1c8ba..f6c2c1c 100644
--- binutils-2.20.1.orig/bfd/elfxx-mips.c
+++ binutils-2.20.1/bfd/elfxx-mips.c
@@ -5688,9 +5688,7 @@ mips_elf_create_dynamic_relocation (bfd *output_bfd,
 
   /* We must now calculate the dynamic symbol table index to use
      in the relocation.  */
-  if (h != NULL
-      && (!h->root.def_regular
-	  || (info->shared && !info->symbolic && !h->root.forced_local)))
+  if (h != NULL && ! SYMBOL_REFERENCES_LOCAL (info, &h->root))
     {
       indx = h->root.dynindx;
       if (SGI_COMPAT (output_bfd))
